File: tools/validate_host_adapter.py
#!/usr/bin/env python3
"""Valida el contrato neutral entre Genesis Core y adaptadores de cuerpo."""

from __future__ import annotations

from copy import deepcopy
import hashlib
import json
from pathlib import Path
import sys
import unicodedata

ROOT = Path(__file__).resolve().parents[1]
VECTORS = ROOT / "conformance" / "host_adapter_vectors.json"

MANIFEST_FIELDS = {
    "schema_version",
    "adapter_id",
    "adapter_version",
    "platform_profile",
    "protocol_versions",
    "verification_state",
    "capabilities",
    "portability",
}
PORTABLE_IDENTITY_FIELDS = {
    "seed_id",
    "seed_root_hash",
    "instance_id",
    "memory",
    "checkpoint",
    "guardian_id",
}
ANCHOR_FIELDS = [
    "protocol_version",
    "seed_root_hash",
    "instance_id",
    "checkpoint_hash",
    "last_event_hash",
    "last_sequence",
    "continuity_status",
    "authority_ledger_head",
]
PORTABILITY_RULES = {
    "neutral_export": (True, "neutral_export_required"),
    "neutral_import": (True, "neutral_import_required"),
    "platform_account_required": (False, "platform_account_required"),
    "private_body_keys_exported": (False, "private_body_key_export_forbidden"),
    "engine_bound_to_identity": (False, "engine_identity_binding_forbidden"),
    "source_deactivation_required": (True, "source_deactivation_required"),
}


class ConformanceError(ValueError):
    """Error estable compartido por los vectores de conformidad."""


def encode_field(value: str) -> bytes:
    if not isinstance(value, str):
        raise ConformanceError("field_must_be_string")
    if unicodedata.normalize("NFC", value) != value:
        raise ConformanceError("text_not_nfc")
    raw = value.encode("utf-8")
    return str(len(raw)).encode("ascii") + b":" + raw + b"\n"


def hash_fields(domain: str, fields: list[str]) -> str:
    preimage = encode_field(domain) + b"".join(encode_field(value) for value in fields)
    return "sha256:" + hashlib.sha256(preimage).hexdigest()


def validate_capabilities(capabilities: list[str], required: list[str]) -> None:
    if len(capabilities) != len(set(capabilities)):
        raise ConformanceError("duplicate_host_capability")
    if capabilities != sorted(capabilities, key=lambda value: value.encode("utf-8")):
        raise ConformanceError("unsorted_host_capabilities")
    missing = set(required) - set(capabilities)
    if missing:
        raise ConformanceError("missing_required_capability")
    if set(capabilities) - set(required):
        raise ConformanceError("unsupported_host_capability")


def validate_portability(portability: dict) -> None:
    missing = set(PORTABILITY_RULES) - set(portability)
    if missing:
        raise ConformanceError("missing_portability_rule")
    if set(portability) - set(PORTABILITY_RULES):
        raise ConformanceError("unexpected_portability_rule")
    for field, (expected, error_code) in PORTABILITY_RULES.items():
        if portability[field] is not expected:
            raise ConformanceError(error_code)


def validate_manifest(manifest: dict, required: list[str]) -> None:
    additional = set(manifest) - MANIFEST_FIELDS
    if additional & PORTABLE_IDENTITY_FIELDS:
        raise ConformanceError("host_manifest_contains_portable_identity")
    if additional:
        raise ConformanceError("unexpected_host_manifest_field")
    if MANIFEST_FIELDS - set(manifest):
        raise ConformanceError("missing_host_manifest_field")
    if manifest["schema_version"] != "genesis.host.capability.manifest.v0.1":
        raise ConformanceError("host_manifest_schema_version_invalid")
    if "genesis.protocol.v0.1" not in manifest["protocol_versions"]:
        raise ConformanceError("protocol_version_not_supported")
    if manifest["verification_state"] not in {
        "declaration_only",
        "simulated",
        "storage_verified",
    }:
        raise ConformanceError("host_verification_state_invalid")
    validate_capabilities(manifest["capabilities"], required)
    validate_portability(manifest["portability"])


def validate_anchor(anchor: dict, forbidden_fields: set[str]) -> None:
    if set(anchor) & forbidden_fields:
        raise ConformanceError("platform_binding_in_portable_anchor")
    if set(anchor) != set(ANCHOR_FIELDS):
        raise ConformanceError("portable_anchor_fields_invalid")


def compute_anchor(vector: dict, forbidden_fields: set[str]) -> str:
    anchor = vector["input"]
    validate_anchor(anchor, forbidden_fields)
    fields = [
        anchor["protocol_version"],
        anchor["seed_root_hash"],
        anchor["instance_id"],
        anchor["checkpoint_hash"],
        anchor["last_event_hash"],
        str(anchor["last_sequence"]),
        anchor["continuity_status"],
        anchor["authority_ledger_head"],
    ]
    return hash_fields(vector["domain"], fields)


def evaluate_rejection(test_case: dict, vectors: dict) -> str | None:
    category = test_case["category"]
    mutation = test_case["input"]
    manifest = deepcopy(vectors["adapters"][0]["manifest"])

    try:
        if category == "required_capabilities":
            capabilities = list(manifest["capabilities"])
            if "remove" in mutation:
                capabilities.remove(mutation["remove"])
            if "duplicate" in mutation:
                capabilities.append(mutation["duplicate"])
            if mutation.get("reverse"):
                capabilities.reverse()
            validate_capabilities(capabilities, vectors["required_capabilities"])
        elif category == "host_manifest_fields":
            for field in mutation["additional_fields"]:
                manifest[field] = "forbidden"
            validate_manifest(manifest, vectors["required_capabilities"])
        elif category == "portability":
            manifest["portability"][mutation["field"]] = mutation["value"]
            validate_manifest(manifest, vectors["required_capabilities"])
        elif category == "portable_anchor_fields":
            anchor = deepcopy(vectors["portable_anchor"]["input"])
            for field in mutation["additional_fields"]:
                anchor[field] = "forbidden"
            validate_anchor(anchor, set(vectors["forbidden_portable_fields"]))
        else:
            raise ConformanceError(f"unknown_host_rejection_category:{category}")
    except ConformanceError as error:
        return str(error)
    return None


def main() -> int:
    vectors = json.loads(VECTORS.read_text(encoding="utf-8"))
    failures: list[str] = []
    if vectors["profile"] != "genesis.host.adapter.v0.1":
        failures.append("host_adapter_profile_invalid")

    required = vectors["required_capabilities"]
    forbidden = vectors["forbidden_portable_fields"]
    try:
        validate_capabilities(required, required)
        if forbidden != sorted(forbidden, key=lambda value: value.encode("utf-8")):
            raise ConformanceError("unsorted_forbidden_portable_fields")
        if len(forbidden) != len(set(forbidden)):
            raise ConformanceError("duplicate_forbidden_portable_field")
    except ConformanceError as error:
        failures.append(str(error))

    expected_profiles = {"android-kotlin", "apple-swift", "windows-dotnet"}
    actual_profiles = {
        adapter["manifest"].get("platform_profile") for adapter in vectors["adapters"]
    }
    if actual_profiles != expected_profiles:
        failures.append("host_fixture_platform_set_invalid")

    expected_anchor = vectors["portable_anchor"]["expected_digest"]
    for adapter in vectors["adapters"]:
        try:
            validate_manifest(adapter["manifest"], required)
            if adapter["manifest"]["verification_state"] != "declaration_only":
                raise ConformanceError("fixture_must_not_claim_storage_verification")
            actual_anchor = compute_anchor(vectors["portable_anchor"], set(forbidden))
            if actual_anchor != expected_anchor:
                raise ConformanceError("portable_anchor_digest_mismatch")
        except ConformanceError as error:
            failures.append(f"{adapter['case_id']}:{error}")

    for test_case in vectors["must_reject"]:
        actual_error = evaluate_rejection(test_case, vectors)
        if actual_error != test_case["expected_error"]:
            failures.append(
                f"{test_case['case_id']}:expected={test_case['expected_error']}:actual={actual_error}"
            )

    if failures:
        for failure in failures:
            print("FAIL", failure)
        return 1

    print("OK portable anchor is identical across Android, Apple, and Windows declarations")
    print(f"OK host capability manifests ({len(vectors['adapters'])})")
    print(f"OK anti-lock-in rejection cases ({len(vectors['must_reject'])})")
    print("NOTE Declaration fixtures are not real platform storage certification.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tools/validate_host_adapter.mjs
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const vectors = JSON.parse(
  fs.readFileSync(path.join(ROOT, "conformance/host_adapter_vectors.json"), "utf8")
);

const MANIFEST_FIELDS = new Set([
  "schema_version",
  "adapter_id",
  "adapter_version",
  "platform_profile",
  "protocol_versions",
  "verification_state",
  "capabilities",
  "portability"
]);
const PORTABLE_IDENTITY_FIELDS = new Set([
  "seed_id",
  "seed_root_hash",
  "instance_id",
  "memory",
  "checkpoint",
  "guardian_id"
]);
const ANCHOR_FIELDS = [
  "protocol_version",
  "seed_root_hash",
  "instance_id",
  "checkpoint_hash",
  "last_event_hash",
  "last_sequence",
  "continuity_status",
  "authority_ledger_head"
];
const PORTABILITY_RULES = new Map([
  ["neutral_export", [true, "neutral_export_required"]],
  ["neutral_import", [true, "neutral_import_required"]],
  ["platform_account_required", [false, "platform_account_required"]],
  ["private_body_keys_exported", [false, "private_body_key_export_forbidden"]],
  ["engine_bound_to_identity", [false, "engine_identity_binding_forbidden"]],
  ["source_deactivation_required", [true, "source_deactivation_required"]]
]);

class ConformanceError extends Error {}

function compareUtf8(left, right) {
  return Buffer.compare(Buffer.from(left, "utf8"), Buffer.from(right, "utf8"));
}

function frame(value) {
  if (typeof value !== "string") throw new ConformanceError("field_must_be_string");
  if (value !== value.normalize("NFC")) throw new ConformanceError("text_not_nfc");
  const bytes = Buffer.from(value, "utf8");
  return Buffer.concat([
    Buffer.from(`${bytes.length}:`, "ascii"),
    bytes,
    Buffer.from("\n", "ascii")
  ]);
}

function hashFields(domain, fields) {
  const preimage = Buffer.concat([frame(domain), ...fields.map((value) => frame(value))]);
  return `sha256:${crypto.createHash("sha256").update(preimage).digest("hex")}`;
}

function validateCapabilities(capabilities, required) {
  if (new Set(capabilities).size !== capabilities.length) {
    throw new ConformanceError("duplicate_host_capability");
  }
  const sorted = [...capabilities].sort(compareUtf8);
  if (capabilities.some((value, index) => value !== sorted[index])) {
    throw new ConformanceError("unsorted_host_capabilities");
  }
  if (required.some((capability) => !capabilities.includes(capability))) {
    throw new ConformanceError("missing_required_capability");
  }
  if (capabilities.some((capability) => !required.includes(capability))) {
    throw new ConformanceError("unsupported_host_capability");
  }
}

function validatePortability(portability) {
  const fields = Object.keys(portability);
  if ([...PORTABILITY_RULES.keys()].some((field) => !fields.includes(field))) {
    throw new ConformanceError("missing_portability_rule");
  }
  if (fields.some((field) => !PORTABILITY_RULES.has(field))) {
    throw new ConformanceError("unexpected_portability_rule");
  }
  for (const [field, [expected, errorCode]] of PORTABILITY_RULES) {
    if (portability[field] !== expected) throw new ConformanceError(errorCode);
  }
}

function validateManifest(manifest, required) {
  const additional = Object.keys(manifest).filter((field) => !MANIFEST_FIELDS.has(field));
  if (additional.some((field) => PORTABLE_IDENTITY_FIELDS.has(field))) {
    throw new ConformanceError("host_manifest_contains_portable_identity");
  }
  if (additional.length > 0) throw new ConformanceError("unexpected_host_manifest_field");
  if ([...MANIFEST_FIELDS].some((field) => !(field in manifest))) {
    throw new ConformanceError("missing_host_manifest_field");
  }
  if (manifest.schema_version !== "genesis.host.capability.manifest.v0.1") {
    throw new ConformanceError("host_manifest_schema_version_invalid");
  }
  if (!manifest.protocol_versions.includes("genesis.protocol.v0.1")) {
    throw new ConformanceError("protocol_version_not_supported");
  }
  if (!["declaration_only", "simulated", "storage_verified"].includes(manifest.verification_state)) {
    throw new ConformanceError("host_verification_state_invalid");
  }
  validateCapabilities(manifest.capabilities, required);
  validatePortability(manifest.portability);
}

function validateAnchor(anchor, forbiddenFields) {
  if (Object.keys(anchor).some((field) => forbiddenFields.has(field))) {
    throw new ConformanceError("platform_binding_in_portable_anchor");
  }
  const actualFields = Object.keys(anchor);
  if (
    actualFields.length !== ANCHOR_FIELDS.length
    || ANCHOR_FIELDS.some((field) => !actualFields.includes(field))
  ) {
    throw new ConformanceError("portable_anchor_fields_invalid");
  }
}

function computeAnchor(vector, forbiddenFields) {
  const anchor = vector.input;
  validateAnchor(anchor, forbiddenFields);
  return hashFields(vector.domain, [
    anchor.protocol_version,
    anchor.seed_root_hash,
    anchor.instance_id,
    anchor.checkpoint_hash,
    anchor.last_event_hash,
    String(anchor.last_sequence),
    anchor.continuity_status,
    anchor.authority_ledger_head
  ]);
}

function evaluateRejection(testCase) {
  const mutation = testCase.input;
  const manifest = structuredClone(vectors.adapters[0].manifest);
  try {
    if (testCase.category === "required_capabilities") {
      const capabilities = [...manifest.capabilities];
      if (mutation.remove) capabilities.splice(capabilities.indexOf(mutation.remove), 1);
      if (mutation.duplicate) capabilities.push(mutation.duplicate);
      if (mutation.reverse) capabilities.reverse();
      validateCapabilities(capabilities, vectors.required_capabilities);
    } else if (testCase.category === "host_manifest_fields") {
      for (const field of mutation.additional_fields) manifest[field] = "forbidden";
      validateManifest(manifest, vectors.required_capabilities);
    } else if (testCase.category === "portability") {
      manifest.portability[mutation.field] = mutation.value;
      validateManifest(manifest, vectors.required_capabilities);
    } else if (testCase.category === "portable_anchor_fields") {
      const anchor = structuredClone(vectors.portable_anchor.input);
      for (const field of mutation.additional_fields) anchor[field] = "forbidden";
      validateAnchor(anchor, new Set(vectors.forbidden_portable_fields));
    } else {
      throw new ConformanceError(`unknown_host_rejection_category:${testCase.category}`);
    }
  } catch (error) {
    if (error instanceof ConformanceError) return error.message;
    throw error;
  }
  return null;
}

function main() {
  const failures = [];
  if (vectors.profile !== "genesis.host.adapter.v0.1") failures.push("host_adapter_profile_invalid");

  try {
    validateCapabilities(vectors.required_capabilities, vectors.required_capabilities);
    const sortedForbidden = [...vectors.forbidden_portable_fields].sort(compareUtf8);
    if (vectors.forbidden_portable_fields.some((value, index) => value !== sortedForbidden[index])) {
      throw new ConformanceError("unsorted_forbidden_portable_fields");
    }
    if (new Set(vectors.forbidden_portable_fields).size !== vectors.forbidden_portable_fields.length) {
      throw new ConformanceError("duplicate_forbidden_portable_field");
    }
  } catch (error) {
    failures.push(error.message);
  }

  const expectedProfiles = new Set(["android-kotlin", "apple-swift", "windows-dotnet"]);
  const actualProfiles = new Set(vectors.adapters.map((adapter) => adapter.manifest.platform_profile));
  if (
    actualProfiles.size !== expectedProfiles.size
    || [...expectedProfiles].some((profile) => !actualProfiles.has(profile))
  ) {
    failures.push("host_fixture_platform_set_invalid");
  }

  const forbiddenFields = new Set(vectors.forbidden_portable_fields);
  for (const adapter of vectors.adapters) {
    try {
      validateManifest(adapter.manifest, vectors.required_capabilities);
      if (adapter.manifest.verification_state !== "declaration_only") {
        throw new ConformanceError("fixture_must_not_claim_storage_verification");
      }
      const actualAnchor = computeAnchor(vectors.portable_anchor, forbiddenFields);
      if (actualAnchor !== vectors.portable_anchor.expected_digest) {
        throw new ConformanceError("portable_anchor_digest_mismatch");
      }
    } catch (error) {
      failures.push(`${adapter.case_id}:${error.message}`);
    }
  }

  for (const testCase of vectors.must_reject) {
    const actualError = evaluateRejection(testCase);
    if (actualError !== testCase.expected_error) {
      failures.push(
        `${testCase.case_id}:expected=${testCase.expected_error}:actual=${actualError}`
      );
    }
  }

  if (failures.length > 0) {
    for (const failure of failures) console.error(`FAIL ${failure}`);
    process.exitCode = 1;
    return;
  }
  console.log("OK portable anchor is identical across Android, Apple, and Windows declarations");
  console.log(`OK host capability manifests (${vectors.adapters.length})`);
  console.log(`OK anti-lock-in rejection cases (${vectors.must_reject.length})`);
  console.log("NOTE Declaration fixtures are not real platform storage certification.");
}

main();
